User defined value types

hrkrshnn opened the pull request:
Closes https://github.com/ethereum/solidity/issues/11531

TODO
- [x] Cleanup all TODOs
- [x] Implement AST import and Export
- [x] Write more tests
- [x] Documentation
- [x] ~NatSpec~?
- [x] ANTLR grammar

hrkrshnn committed:
Added AST node UserDefinedValueType

hrkrshnn committed:
Implemented parsing for UserDefinedValueType

Also added parsing tests.

hrkrshnn committed:
Types for UserDefinedValueType

hrkrshnn committed:
Implemented codegen for UserDefinedValueType

chriseth commented on libsolidity/analysis/DeclarationContainer.cpp:
```diff
@@ -50,7 +50,8 @@ Declaration const* DeclarationContainer::conflictingDeclaration(
 		dynamic_cast<FunctionDefinition const*>(&_declaration) ||
 		dynamic_cast<EventDefinition const*>(&_declaration) ||
-		dynamic_cast<MagicVariableDeclaration const*>(&_declaration)
+		dynamic_cast<MagicVariableDeclaration const*>(&_declaration) ||
```
Isn't this the list of symbols that can be overloaded? I'm not sure `UserDefinedValueType` belongs here.

chriseth commented on libsolidity/analysis/DeclarationTypeChecker.cpp:
```diff
@@ -158,6 +158,20 @@ void DeclarationTypeChecker::endVisit(UserDefinedTypeName const& _typeName)
+	{
+		TypeName const* typeName = userDefinedValueType->typeName();
+		solAssert(typeName, "");
+		if (!dynamic_cast<ElementaryTypeName const*>(typeName))
```
Don't we also want to restrict these to value types? (`string` and `bytes` are also elementary type names)

hrkrshnn replied:
This is done in the TypeChecker.

chriseth replied:
But we could do it already here, can't we? Shouldn't we do it as early as possible? Otherwise we might get into trouble if we have a forward reference and the type checker somehow assumes that the underlying type is a value type.

hrkrshnn replied:
Sure, I can move it here.

chriseth commented on libsolidity/analysis/DeclarationTypeChecker.cpp:
```diff
@@ -158,6 +158,20 @@ void DeclarationTypeChecker::endVisit(UserDefinedTypeName const& _typeName)
+				_typeName.location(),
+				"The type has to be an elementary value type."
+			);
+		Type const* actualType = userDefinedValueType->typeName()->annotation().type;
```
"underlyingType"?

chriseth commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -1700,21 +1700,33 @@ Type const* TypeChecker::typeCheckTypeConversionAndRetrieveReturnType(
-		);
+		m_errorReporter
+			.typeError(5153_error, _functionCall.location(), "Type conversion cannot allow named arguments.");
+	else if (auto userDefinedValueTypeType = dynamic_cast<UserDefinedValueTypeType const*>(expressionType))
```
Do we need a special case here? Isn't this handled by `isExplicitlyConvertibelTo`?

chriseth commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -2421,6 +2433,15 @@ bool TypeChecker::visit(FunctionCall const& _functionCall)
 		break;
 	}
 
+	case Type::Category::UserDefinedValueType:
```
The conversion should be handled by the case above (`TypeType`).

hrkrshnn committed:
DeclarationTypeChecker for UserDefinedValueTypes

- Checks for repetitions.
- Assigns types.
- Checks if the 'actual type' is an elementary type name.
- Checks if the 'actual type' is a value type.

Also added tests.

chriseth commented on libsolidity/parsing/Parser.h:
```diff
@@ -108,6 +108,7 @@ class Parser: public langutil::ParserBase
 	ASTPointer<Identifier> parseIdentifier();
 	ASTPointer<Identifier> parseIdentifierOrAddress();
 	ASTPointer<UserDefinedTypeName> parseUserDefinedTypeName();
+	ASTPointer<UserDefinedValueType> parseUserDefinedValueType();
```
Shouldn't this be placed closer to StructDefinition and EnumDefinition?

chriseth replied:
Also maybe `parseUserDefinedValueTypeDefinition`?

chriseth commented on libsolidity/parsing/Parser.cpp:
```diff
@@ -386,6 +392,12 @@ ASTPointer<ContractDefinition> Parser::parseContractDefinition()
 			else if (currentTokenValue == Token::Using)
 				subNodes.push_back(parseUsingDirective());
+			else if (
+				currentTokenValue == Token::Type &&
```
Isn't `type` enough of a trigger?

chriseth commented on libsolidity/ast/Types.h:
```diff
@@ -1454,6 +1455,54 @@ class TypeType: public Type
+{
+public:
+	explicit UserDefinedValueTypeType(
+		Type const& _actualType,
```
`_underlyingType`?

chriseth commented on libsolidity/ast/Types.h:
```diff
@@ -1454,6 +1455,54 @@ class TypeType: public Type
+/**
+ * The type of a UserDefinedValueType.
+ */
+class UserDefinedValueTypeType: public Type
```
Why TypeType?
```
type NewType is int;
NewType a = NewType(2);
a;
```

In the above, the type of `a` (the one in isolation) is `UserDefinedValueType(int)`, while the type of `NewType` (as sub-expression of `NewType(2)` is `TypeType(UserDefinedValueType(int))`

If you treat the newly introduced type class more like the `Enum` or `Struct` class, then most of the other things should be automatic.

hrkrshnn committed:
Semantic tests for User Defined Value types

hrkrshnn committed:
Implemented ASTJson import export for UserDefinedValueType

Also added a test.

hrkrshnn committed:
Documentation and Changelog

chriseth commented on docs/types/value-types.rst:
```diff
@@ -617,6 +617,49 @@ subsequent unsigned integer values starting from ``0``.
+
+A user defined value type allows creating a zero cost abstraction over a value type. This is similar
+to an alias, but with stricter type requirements.
+
```
Maybe insert:

A user-defined value type is defined using ``type C is V``, where `C` is the name of the newly introduced type and `V` has to be a built-in value type.

chriseth commented on docs/types/value-types.rst:
```diff
@@ -617,6 +617,49 @@ subsequent unsigned integer values starting from ``0``.
+For a custom type ``C``, defined by ``type C is V;`` where ``V`` is the underlying value type, the
+following type requirements apply:
+
+- Implicit conversions of ``C`` are only allowed with ``C``.
```
I would not list this, since it is not really a conversion

chriseth commented on docs/types/value-types.rst:
```diff
@@ -617,6 +617,49 @@ subsequent unsigned integer values starting from ``0``.
+following type requirements apply:
+
+- Implicit conversions of ``C`` are only allowed with ``C``.
+- Explicit conversions of ``C`` are only allowed with itself and the underlying type, i.e., ``C``
```
Same here

chriseth commented on docs/types/value-types.rst:
```diff
@@ -617,6 +617,49 @@ subsequent unsigned integer values starting from ``0``.
+- Implicit conversions of ``C`` are only allowed with ``C``.
+- Explicit conversions of ``C`` are only allowed with itself and the underlying type, i.e., ``C``
+  and ``V`` respectively.
+- The type ``C`` does not have any operators or bound member functions. In particular, the operator
+  ``==`` is disallowed. One can use operators or bound member functions by explicitly converting it
```
```suggestion
- The type ``C`` does not have any operators or bound member functions. In particular, even the operator
  ``==`` is not defined. One can use operators or bound member functions by explicitly converting it
```

chriseth commented on docs/types/value-types.rst:
```diff
@@ -617,6 +617,49 @@ subsequent unsigned integer values starting from ``0``.
+  ``==`` is disallowed. One can use operators or bound member functions by explicitly converting it
+  to the underlying type.
+
+The following example demonstrates a custom type ``MyAddress`` which is an abstraction over the
```
Shouldn't we use an example that also shows the usefulness of this concept?

chriseth commented on libsolidity/ast/AST.h:
```diff
@@ -726,6 +726,37 @@ class EnumValue: public Declaration
+ * User defined value types, i.e., custom types, for example, `type MyInt is int`. Allows creating a
+ * zero cost abstraction over value type with stricter type requirements.
+ */
+class UserDefinedValueType: public Declaration
```
```suggestion
class UserDefinedValueTypeDefinition: public Declaration
```

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -410,6 +420,8 @@ BoolResult AddressType::isExplicitlyConvertibleTo(Type const& _convertTo) const
 {
 	if ((_convertTo.category() == category()) || isImplicitlyConvertibleTo(_convertTo))
 		return true;
+	else if (toUserDefinedValueType(_convertTo, *this))
```
Maybe implement this in the Type class?

chriseth replied:
Or in the UserDefinedValueType class?

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -3780,6 +3804,31 @@ BoolResult TypeType::isExplicitlyConvertibleTo(Type const& _convertTo) const
 
+string UserDefinedValueTypeType::richIdentifier() const
+{
+	return "t_userDefinedValueType" + identifierList(&underlyingType());
```
```suggestion
	return "t_userDefinedValueType" + parenthesizedIdentifier(underlyingType());
```

hrkrshnn commented on test/libsolidity/syntaxTests/userDefinedValueType/all_value_types.sol:
```diff
@@ -0,0 +1,21 @@
+type MyFunctionWithCustomParam is function(MyAddress) returns (MyInt);
+
+
+// TODO should these be disallowed?
```
@chriseth Should these be disallowed?

chriseth replied:
Yes, I would disallow them, at least for now.

hrkrshnn commented on test/libsolidity/syntaxTests/userDefinedValueType/explicit_conversions.sol:
```diff
@@ -0,0 +1,10 @@
+type MyAddress is address;
+
+function f() pure {
+    // TODO Should we allow explicit conversions from literals?
```
I think yes. This is similar to how newtype semantics work in haskell and rust (?)

chriseth replied:
I'm not too sure about it - wouldn't it create ambiguities? We have to at least check that implicit conversion from the literal to the underlying type is possible.

chriseth commented on libsolidity/ast/Types.h:
```diff
@@ -176,7 +176,7 @@ class Type
 		FixedBytes, Contract, Struct, Function, Enum, Tuple,
 		Mapping, TypeType, Modifier, Magic, Module,
-		InaccessibleDynamic
+		InaccessibleDynamic, UserDefinedValueType
```
Maybe move it closer to struct and enum?

chriseth commented on libsolidity/ast/Types.h:
```diff
@@ -1076,6 +1077,56 @@ class EnumType: public Type
+
+	Category category() const override { return Category::UserDefinedValueType; }
+	Type const& underlyingType() const { return m_underlyingType; }
+	UserDefinedValueTypeDefinition const& userDefinedValueType() const { return m_userDefinedValueType; }
```
Do you need this? If yes maybe call the function `definition()`?

chriseth commented on libsolidity/ast/Types.h:
```diff
@@ -1076,6 +1077,56 @@ class EnumType: public Type
+		solAssert(m_underlyingType.isValueType(), "");
+		return true;
+	}
+	bool nameable() const override { return m_underlyingType.nameable(); }
```
I think we should assert that the underlying type is nameable.

chriseth commented on libsolidity/ast/Types.h:
```diff
@@ -1076,6 +1077,56 @@ class EnumType: public Type
+	}
+	bool nameable() const override { return m_underlyingType.nameable(); }
+
+	std::string toString(bool _short) const override { return "UserDefinedValueType(" + m_underlyingType.toString(_short) + ")"; }
```
Shouldn't this also include the name of the type? (as we do for struct)

chriseth commented on libsolidity/ast/Types.h:
```diff
@@ -1076,6 +1077,56 @@ class EnumType: public Type
+	std::string toString(bool _short) const override { return "UserDefinedValueType(" + m_underlyingType.toString(_short) + ")"; }
+
+	BoolResult isExplicitlyConvertibleTo(Type const& _convertTo) const override;
+	///	A helper function to reason about UserDefinedValueType to ValueType explicit conversion
```
```suggestion
	/// helper function to reason about UserDefinedValueType to ValueType explicit conversion
```

chriseth commented on libsolidity/ast/Types.h:
```diff
@@ -1076,6 +1077,56 @@ class EnumType: public Type
+	static bool convertTo(Type const& _userDefined, Type const& _valueType);
+protected:
+	std::vector<std::tuple<std::string, Type const*>> makeStackItems() const override;
+
```
I think we should also overwrite the following two functions:
 - canonicalName (assert false)
 - signatureInExternalFunction (assert false)

chriseth commented on libsolidity/codegen/CompilerUtils.cpp:
```diff
@@ -772,6 +772,33 @@ void CompilerUtils::convertType(
+		return
+			convertType(
+				_targetType,
+				dynamic_cast<UserDefinedValueTypeType const&>(_typeOnStack).underlyingType(),
```
shouldn't this be
```suggestion
				dynamic_cast<UserDefinedValueTypeType const&>(_typeOnStack).underlyingType(),
				_targetType,
```
?

chriseth commented on libsolidity/codegen/YulUtilFunctions.cpp:
```diff
@@ -3431,6 +3443,15 @@ string YulUtilFunctions::conversionFunction(Type const& _from, Type const& _to)
 			body = "converted := value";
 			break;
 		}
+		case Type::Category::UserDefinedValueType:
```
Can this still happen?

chriseth commented on libsolidity/codegen/ir/IRGeneratorForStatements.cpp:
```diff
@@ -869,8 +869,10 @@ void IRGeneratorForStatements::endVisit(FunctionCall const& _functionCall)
 		solAssert(
-			_functionCall.expression().annotation().type->category() == Type::Category::TypeType,
+			category == Type::Category::TypeType ||
+			category == Type::Category::UserDefinedValueType,
```
I don't think this special case is needed if the type checker properly assigns a typetype to `MyType` in the expression `MyType(t)`

chriseth commented on libsolidity/parsing/Parser.cpp:
```diff
@@ -134,6 +134,10 @@ ASTPointer<SourceUnit> Parser::parse(CharStream& _charStream)
 					nodes.push_back(parseVariableDeclaration(options));
 					expectToken(Token::Semicolon);
 				}
+				else if (
+					m_scanner->currentToken() == Token::Type
+				)
```
```suggestion
				else if (m_scanner->currentToken() == Token::Type)
```

chriseth replied:
Actually why not move this to a `case`? The part here should only be the fallback / workaround case

chriseth commented on libsolidity/parsing/Parser.cpp:
```diff
@@ -386,6 +390,10 @@ ASTPointer<ContractDefinition> Parser::parseContractDefinition()
 				subNodes.push_back(parseEventDefinition());
 			else if (currentTokenValue == Token::Using)
 				subNodes.push_back(parseUsingDirective());
+			else if (
+				currentTokenValue == Token::Type
+			)
```
```suggestion
			else if (currentTokenValue == Token::Type)
```

hrkrshnn committed:
Updated ANTLR grammar to include user defined value types.

hrkrshnn commented on test/libsolidity/semanticTests/userDefinedValueType/erc20.sol:
```diff
@@ -0,0 +1,146 @@
+// transfer(address,uint256): 2, 2 -> FAILURE, hex"4e487b71", 0x11
+// gas irOptimized: 24052
+// gas legacy: 24492
+// gas legacyOptimized: 24074
```
For a gas diff, compared to our other erc20 test: https://gist.github.com/hrkrshnn/f7e6e69edfe62d5f94648a7b4f4abb26

I think the difference is because of inlining for the library based safemath.

chriseth commented on docs/abi-spec.rst:
```diff
@@ -89,6 +89,10 @@ The following non-fixed-size types exist:
 
 - ``<type>[]``: a variable-length array of elements of the given type.
 
+For :ref:`user defined value types<user-defined-value-types>`:
```
I think this belongs into the "Mapping Solidity to ABI types".

chriseth commented on docs/grammar/SolidityParser.g4:
```diff
@@ -247,6 +249,11 @@ structMember: type=typeName name=identifier Semicolon;
+ * Definition of a user defined value type. Can occur at top-level within a source unit or within a contract, library or interface.
+ */
+userDefinedValueTypeDefinition:
+	Type name=identifier Is elementaryTypeName[true] Semicolon;
```
```suggestion
	Type name=identifier Is elementaryTypeName[false] Semicolon;
```
I would not allow `address payable` here since it serves no purpose and makes the grammar more complicated.

hrkrshnn replied:
Do you want to disallow `type MyPayableAddress is address payable`? This is currently allowed. I think there is no good reason to disallow it.

ekpyron replied:
I'd say it should be allowed.

chriseth replied:
Ok!

chriseth commented on docs/types/value-types.rst:
```diff
@@ -617,6 +617,51 @@ subsequent unsigned integer values starting from ``0``.
+
+.. _user-defined-value-types:
+
+User Defined Value Types
```
Is it "user defined" or "user-defined"?

chriseth commented on libsolidity/analysis/DeclarationTypeChecker.cpp:
```diff
@@ -140,6 +140,36 @@ bool DeclarationTypeChecker::visit(StructDefinition const& _struct)
 	return false;
 }
 
+bool DeclarationTypeChecker::visit(UserDefinedValueTypeDefinition const& _userDefined)
```
Why can't we do this both in `endVisit`?

chriseth replied:
Ok, I see why you did it this way: In `DeclarationTypeChecker::endVisit(UserDefinedTypeName`, the underlying type needs to be available, but I think this is a more serious problem and cannot be solved by changing the visitation order: We can also have forward references to other user-defined types at that point. I think the solution is the same as for struct types: Change the type constructor so that it only takes the AST definition, but do not analyze the type further.

chriseth replied:
Indeed, the following code still triggers the assertion:
```
function f(MyIntB x) pure {}
type MyIntB is MyIntB;
```

chriseth commented on test/libsolidity/semanticTests/userDefinedValueType/calldata.sol:
```diff
@@ -0,0 +1,72 @@
+// test_f() -> true
+// gas irOptimized: 96989275
+// gas legacy: 96989774
+// gas legacyOptimized: 96989223
```
This is weird - it got almost 100 gas more expensive after my changes.

ekpyron replied:
If we can't get around cost otherwise, I'd say we just add the functions to the type instead.
I mean - they should be there anyways - otherwise you'll have to write these wrapping functions all the time for each type. And then we could actually more easily also allow function types, etc.

hrkrshnn committed:
TypeChecker: added checks for `wrap` and `unwrap`

For user defined value types and also added a test case.

hrkrshnn committed:
Syntax checks for user defined value types

ekpyron commented on test/libsolidity/syntaxTests/userDefinedValueType/recursive_err.sol:
```diff
@@ -0,0 +1,4 @@
+type MyInt1 is MyInt2;
+type MyInt2 is MyInt1;
+// ----
+// TypeError 8657: (15-21): The underlying type for a user defined value type has to be an elementary value type.
```
This is a bad restriction. We can of course keep this for now, but we *do* plan to lift this eventually, right :-)?

ekpyron commented on test/libsolidity/syntaxTests/userDefinedValueType/non_value_type_function_err.sol:
```diff
@@ -0,0 +1,3 @@
+type MyFunction is function(uint) returns (uint);
+// ----
+// TypeError 8657: (19-49): The underlying type for a user defined value type has to be an elementary value type.
```
Since we now generate the wrapping and unwrapping functions internally, is there any reason for this not to be allowed? But also fine to keep disallowed for now.

ekpyron commented on libsolidity/codegen/ir/IRGeneratorForStatements.cpp:
```diff
@@ -1051,6 +1051,20 @@ void IRGeneratorForStatements::endVisit(FunctionCall const& _functionCall)
+	{
+		solAssert(arguments.size() == 1, "");
+		FunctionType::Kind kind = functionType->kind();
+		auto& userDefinedType = kind == FunctionType::Kind::Wrap ?
+			dynamic_cast<UserDefinedValueType const&>(type(_functionCall)) :
+			dynamic_cast<UserDefinedValueType const&>(type(*arguments.at(0)));
+		if (kind == FunctionType::Kind::Wrap)
+			solAssert(type(*arguments.at(0)).isImplicitlyConvertibleTo(userDefinedType.underlyingType()), "");
```
Not sure this is necessary, but if we want it, then

```suggestion
		if (kind == FunctionType::Kind::Wrap)
			solAssert(type(*arguments.at(0)).isImplicitlyConvertibleTo(
				dynamic_cast<UserDefinedValueType const&>(type(_functionCall))
			);
```
would be enough.

ekpyron commented on libsolidity/codegen/YulUtilFunctions.cpp:
```diff
@@ -3696,6 +3706,9 @@ string YulUtilFunctions::arrayConversionFunction(ArrayType const& _from, ArrayTy
 
 string YulUtilFunctions::cleanupFunction(Type const& _type)
 {
+	if (auto userDefinedValueType = dynamic_cast<UserDefinedValueType const*>(&_type))
+		return cleanupFunction(userDefinedValueType->underlyingType());
```
It'd probably be more consistent to actually generate a function that's a thin wrapper around the underlying cleanup function similarly to the other proxy cases below.

ekpyron commented on libsolidity/codegen/ExpressionCompiler.cpp:
```diff
@@ -957,6 +957,23 @@ bool ExpressionCompiler::visit(FunctionCall const& _functionCall)
+			Type const* argumentType = arguments.at(0)->annotation().type;
+			solAssert(argumentType, "");
+			FunctionType::Kind kind = functionType->kind();
+			auto userDefinedType = kind == FunctionType::Kind::Wrap ?
+				dynamic_cast<UserDefinedValueType const*>(_functionCall.annotation().type) :
+				dynamic_cast<UserDefinedValueType const*>(argumentType);
+			solAssert(userDefinedType, "");
+			if (kind == FunctionType::Kind::Wrap)
+				solAssert(argumentType->isImplicitlyConvertibleTo(userDefinedType->underlyingType()), "");
```
```suggestion
			if (kind == FunctionType::Kind::Wrap)
			{
				auto userDefinedType = dynamic_cast<UserDefinedValueType const*>(_functionCall.annotation().type);
				solAssert(userDefinedType, "");
				solAssert(argumentType->isImplicitlyConvertibleTo(userDefinedType->underlyingType()), "");
			}
```

ekpyron commented:
How about dropping the ``Value`` part and calling these things just plain ``UserDefinedType`` everywhere in the sources, docs and messages and rather note that this is *currently* restricted to non-function value types :-)?

ekpyron commented on test/libsolidity/ABIJson/user_defined_value_type.sol:
```diff
@@ -0,0 +1,173 @@
+//     "inputs":
+//     [
+//       {
+//         "internalType": "user defined type MyUInt8",
```
Ah, nice, so here the ``value`` thing is already skipped, that's good.

ekpyron commented on test/libsolidity/ASTJSON/userDefinedValueType.json:
```diff
@@ -0,0 +1,264 @@
+      "id": 2,
+      "name": "MyAddress",
+      "nameLocation": "5:9:1",
+      "nodeType": "UserDefinedValueTypeDefinition",
```
```suggestion
      "nodeType": "UserDefinedValueTypeDefinition",
```
As for names: This I would rather have say ``UserDefinedTypeDefinition`` or even just plainly ``TypeDefinition``.

chriseth commented on docs/abi-spec.rst:
```diff
@@ -113,6 +113,9 @@ them.
 +-------------------------------+-----------------------------------------------------------------------------+
 |:ref:`enum<enums>`             |``uint8``                                                                    |
 +-------------------------------+-----------------------------------------------------------------------------+
+|:ref:`user defined value types |same as its underlying value type                                            |
```
"same as" sounds like it is not exactly the underlying value type

hrkrshnn replied:
Changed it to "exactly as its underlying value type"

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -3754,6 +3786,34 @@ MemberList::MemberMap TypeType::nativeMembers(ASTNode const* _currentScope) cons
 	}
+	else if (m_actualType->category() == Category::UserDefinedValueType)
+	{
+		auto& userDefined = dynamic_cast<UserDefinedValueType const&>(*m_actualType);
```
```suggestion
		auto const& userDefined = dynamic_cast<UserDefinedValueType const&>(*m_actualType);
```

hrkrshnn replied:
But the `const` here is redundant, isn't it?

chriseth replied:
It is, but may be nice to mention it - we usually do.

chriseth commented on libsolidity/codegen/ExpressionCompiler.cpp:
```diff
@@ -957,6 +957,23 @@ bool ExpressionCompiler::visit(FunctionCall const& _functionCall)
+			if (kind == FunctionType::Kind::Wrap)
+				solAssert(argumentType->isImplicitlyConvertibleTo(userDefinedType->underlyingType()), "");
+
+			acceptAndConvert(*arguments[0], *function.parameterTypes()[0], true);
```
Do we really need to do cleanup?

chriseth replied:
Can you assert that the implicit conversion is possible, please?

hrkrshnn replied:
I've removed the cleanup and added one more assert.

chriseth commented on libsolidity/codegen/ExpressionCompiler.cpp:
```diff
@@ -957,6 +957,23 @@ bool ExpressionCompiler::visit(FunctionCall const& _functionCall)
+			Type const* argumentType = arguments.at(0)->annotation().type;
+			solAssert(argumentType, "");
+			FunctionType::Kind kind = functionType->kind();
+			auto userDefinedType = kind == FunctionType::Kind::Wrap ?
```
Wouldn't it be easier and less confusing to retrieve this from `_functionCall.expression`? Maybe not.

ekpyron replied:
On that note test cases for stuff like ``(MyType).wrap`` and the like would be good.

chriseth commented on docs/abi-spec.rst:
```diff
@@ -113,6 +113,9 @@ them.
 +-------------------------------+-----------------------------------------------------------------------------+
 |:ref:`enum<enums>`             |``uint8``                                                                    |
 +-------------------------------+-----------------------------------------------------------------------------+
+|:ref:`user defined value types |exactly as its underlying value type                                         |
```
Sorry, the "exactly" was not meant literally. The text above table says "on the right column the ABI types that represent them", so it should be enough to say "its underlying value type" or "the underlying value type".

hrkrshnn replied:
Fixed

chriseth commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -2481,6 +2481,13 @@ bool TypeChecker::visit(FunctionCall const& _functionCall)
+		case FunctionType::Kind::Wrap:
+		case FunctionType::Kind::Unwrap:
+		{
+			typeCheckFunctionGeneralChecks(_functionCall, functionType);
```
Ah nice!

chriseth commented on libsolidity/codegen/ExpressionCompiler.cpp:
```diff
@@ -957,6 +957,34 @@ bool ExpressionCompiler::visit(FunctionCall const& _functionCall)
+			FunctionType::Kind kind = functionType->kind();
+			if (kind == FunctionType::Kind::Wrap)
+			{
+				solAssert(
+					argumentType->isImplicitlyConvertibleTo(
+					dynamic_cast<UserDefinedValueType const&>(*functionCallType).underlyingType()),
+					""
+				);
```
```suggestion
				solAssert(
					argumentType->isImplicitlyConvertibleTo(
						dynamic_cast<UserDefinedValueType const&>(*functionCallType).underlyingType()
					),
					""
				);
```

chriseth commented on libsolidity/codegen/ir/IRGeneratorForStatements.cpp:
```diff
@@ -1051,6 +1051,23 @@ void IRGeneratorForStatements::endVisit(FunctionCall const& _functionCall)
+		if (kind == FunctionType::Kind::Wrap)
+			solAssert(
+				type(*arguments.at(0)).isImplicitlyConvertibleTo(
+					dynamic_cast<UserDefinedValueType const&>(type(_functionCall)).underlyingType()),
```
```suggestion
					dynamic_cast<UserDefinedValueType const&>(type(_functionCall)).underlyingType()
				),
```

chriseth commented on libsolidity/codegen/ir/IRGeneratorForStatements.cpp:
```diff
@@ -1051,6 +1051,23 @@ void IRGeneratorForStatements::endVisit(FunctionCall const& _functionCall)
+		else
+			solAssert(type(*arguments.at(0)).category() == Type::Category::UserDefinedValueType, "");
+
+		define(_functionCall, *arguments.at(0));
```
This actually skips one conversion - but would be fine for me, since the conversion is a no-op (and I hope it stays like that).

ekpyron commented on libsolidity/ast/AST.h:
```diff
@@ -726,6 +726,37 @@ class EnumValue: public Declaration
+ * User defined value types, i.e., custom types, for example, `type MyInt is int`. Allows creating a
+ * zero cost abstraction over value type with stricter type requirements.
+ */
+class UserDefinedValueTypeDefinition: public Declaration
```
I *still* think this is a *really* bad name for the node :-D - but well.

ekpyron commented on test/libsolidity/syntaxTests/userDefinedValueType/in_parenthesis.sol:
```diff
@@ -0,0 +1,7 @@
+type MyInt is int;
+function f() pure {
+    (MyInt).wrap;
+    (MyInt).wrap(5);
+    (MyInt).unwrap;
+    (MyInt).unwrap(MyInt.wrap(5));
```
Syntactically this is really not interesting - this should be a semantics test!

hrkrshnn replied:
I can move this.

hrkrshnn replied:
Did it here: https://github.com/ethereum/solidity/pull/11914/

fulldecent reviewed (commented):
The fixed math library claims the division function is to the "nearest integer" but actually it rounds down.

hrkrshnn commented:
@fulldecent Thanks for noticing. Will update the comment.

fulldecent commented:
Thank you! Looks good now
